crates/consensus/src/lib.rs

```diff
@@ -962,6 +1056,79 @@ pub trait ValidatorSetProvider<A> {
     fn get_validator_set(&self, height: u64) -> Result<ValidatorSet<A>, anyhow::Error>;
 }
 
+/// A trait for selecting the proposer for a given height and round.
+///
+/// This trait allows consumers to provide custom proposer selection logic
+/// instead of using the default round-robin approach. This is useful for
+/// implementing more sophisticated proposer selection algorithms like
+/// VRF-based selection, weighted selection, or other custom logic.
+///
+/// ## Example Implementation
+///
+/// ```rust
+/// use pathfinder_consensus::*;
+///
+/// struct RoundRobinProposerSelector;
+///
+/// impl<A: ValidatorAddress> ProposerSelector<A> for RoundRobinProposerSelector {
+///     fn select_proposer<'a>(
+///         &self,
+///         validator_set: &'a ValidatorSet<A>,
+///         height: u64,
+///         round: u32,
+///     ) -> &'a Validator<A> {
+///         let index = round as usize % validator_set.count();
+///         &validator_set.validators[index]
+///     }
+/// }
+/// ```
+pub trait ProposerSelector<A: ValidatorAddress>: Clone + Send + Sync {
+    /// Select the proposer for the given height and round.
+    ///
+    /// ## Arguments
+    ///
+    /// - `validator_set`: The set of validators eligible to propose
+    /// - `height`: The blockchain height
+    /// - `round`: The consensus round number
+    ///
+    /// ## Returns
+    ///
+    /// Returns a reference to the selected validator who should propose
+    /// for the given height and round.
+    fn select_proposer<'a>(
+        &self,
+        validator_set: &'a ValidatorSet<A>,
+        height: u64,
+        round: u32,
+    ) -> &'a Validator<A>;
+}
+
+/// A default proposer selector that uses round-robin selection.
+///
+/// This is the default proposer selection algorithm that selects proposers
+/// in a round-robin fashion based on the round number modulo the number of
+/// validators.
+#[derive(Clone, Default)]
+pub struct RoundRobinProposerSelector;
+
+impl<A: ValidatorAddress> ProposerSelector<A> for RoundRobinProposerSelector {
+    fn select_proposer<'a>(
+        &self,
+        validator_set: &'a ValidatorSet<A>,
+        _height: u64,
```

Comment:
shouldn't the proposer change by height, too?

Reply:
I've kept the [same logic](https://github.com/eqlabs/pathfinder/blob/main/crates/consensus/src/internal/malachite.rs#L403) we've had until now to make things backwards-compatible